### PR TITLE
crypto.keccak.State: don't unconditionally permute after a squeeze()

### LINKED ISSUE #22019: std.crypto.core.keccak: do not overpermute the sponge
### Zig Version

0.12.0+efe06c5f3

### Steps to Reproduce and Observed Behavior

The following code:

```zig
const std = @import("std");
const keccak = std.crypto.core.keccak;
const State = keccak.State;

pub fn main() !void {
    var k: State(1600, 512, 24) = .{ .delim = 0x01 };
    k.absorb(&[_]u8 { 0x13 } ** 136); // not important
    var out: [136]u8 = undefined;
    k.squeeze(out[0..]);

    std.debug.print("{any}", .{k.st.st});
}
```

results in

```
{ 7233668778566134865, 11592713930134003118, 15373517831437853021, 15608250416844938208, 11330810810757603158, 6337136767010511286, 1240513802066837974, 4881489601663522637, 1927467063577625972, 17630746601708290681, 12181661910834865096, 13184932669851834457, 11304855261261663641, 8294672726897056558, 8839467195840986991, 7083940663377158003, 11394419039461769406, 8921936926094339996, 2284247437489028585, 13015763828452273947, 15409054831738269266, 10463282502587619091, 600522821999583161, 7632801731005224094, 8117278901483848959 }
```

### Expected Behavior

The result of the above *should* be

```
{ 3064831273020639228, 13512604793548462962, 12805899729923228041, 6119062306464645338, 4536591997206582897, 14771343959304370689, 2013378616438251871, 14173905344357195255, 4698469717602941177, 5222380363174343544, 13163819436985973242, 4908270899219160089, 12120744815160992576, 13809031795350906813, 533294723529621938, 14014335485794523625, 17720142420523044394, 18278950936426394526, 15638265930130015358, 14236447863273458192, 16962296917699274481, 10248514880263995001, 1834222953400603767, 2439774991396631628, 8370602107368013381 }
```

The reason for this is that the `squeeze` function permutes the state even when the requested output length is not greater than the rate.

1. The current implementation does not follow the specification (e.g. Algorithm 4 in [CSF](https://keccak.team/files/CSF-0.1.pdf)). Which makes it hard to use the `squeeze` as a primitive.
2. The current implementation will permute the state one  more time than what is needed in for example `Sha3` since it permutes the state after the final `extractBytes` but never uses the permuted state.

Caveats; there may be some update to the specification of which I am unaware, but it is notable that also [XKCP the reference implementation in Python](https://github.com/XKCP/XKCP/blob/master/Standalone/CompactFIPS202/Python/CompactFIPS202.py) does not conform with the current Zig standard libray implementation. They do however not put the squeeze part in a separate function.

Finally; changing the [line 277](https://github.com/ziglang/zig/blob/master/lib/std/crypto/keccak_p.zig#L277) to include a condition (for example `if (i + rate < out.len)`) does not seem to change the result of any of the unit tests (in particular all the `Sha3` unit tests all still pass). It (presumably) only makes it faster.

### PR DESCRIPTION
**Pull request by jedisct1**
Now, squeeze() behaves like absorb()

Namely,

```
squeeze(x[0..t]);
squeeze(x[t..n)); with t <= n
```

becomes equivalent to `squeeze(x[0..n])`.

Fixes #22019

### CODE REVIEW

**Review by krutimyltan2 — COMMENTED** [review]
This change scares me.

The same `offset` is used for both absorbing and squeezing which makes it possible to jump between absorbing and squeezing without permuting the state. I have recently been informed that this may be considered bad practice.

Furthermore, consider the following example written as a test case
```zig
test "spooky behaviour" {
    const st: State(1600, 512, 24) = .{ .delim = 0x01 };

    const in0: [136 / 2]u8 = [_]u8 { 55 } ** (136 / 2);
    var in1: [136 / 2]u8 = [_]u8 { 77 } ** (136 / 2);
    in1[0] = 55;
    var out_pre: [135]u8 = undefined;
    var out0: [136 / 2 - 1]u8 = undefined;
    var out1: [136 / 2 - 1]u8 = undefined;

    var st0 = st;
    st0.squeeze(out_pre[0..]); // squeeze one byte short of block
    st0.absorb(in0[0..]); // absorb half a block
    st0.squeeze(out0[0..]); // squeeze almost a block

    var st1 = st;
    st1.squeeze(out_pre[0..]); // squeeze one byte short of block
    st1.absorb(in1[0..]); // absorb half a block
    st1.squeeze(out1[0..]); // squeeze almost a block

    // although in0 and in1 only have one byte in common, the
    // outputs out0 and out1 are equal. intended?
    try std.testing.expectEqualSlices(u8, &out0, &out1);
}
```

The squeeze does not have to depend on everything that has been absorbed (or am I mistaken?). I don't like that.

**Comment by jedisct1** [comment]
> The same offset is used for both absorbing and squeezing which makes it possible to jump between absorbing and squeezing without permuting the state.

Using a shared offset variable makes reading and writing the same location mutually exclusive, unless there's a permutation in-between.

We can still add consistency checks to forbid direct transitions from absorb to squeeze, that would be only enabled in `Debug` mode.

But of course if a stated is cloned, and the different instances are used in different ways, everything falls apart, even from a protocol perspective.

**Comment by jedisct1** [comment]
Done. In debug mode, transitions are now tracked to prevent insecure ones.

**Comment by jedisct1** [comment]
The transition checks immediately spotted an inconsistency in the squeeze test, so I'm now convinced that they're indeed good to have.

File: lib/std/crypto/keccak_p.zig [hunk 1]
```diff
@@ -294,12 +294,17 @@ pub fn State(comptime f: u11, comptime capacity: u11, comptime rounds: u5) type
 
         /// Permute the state
         pub fn permute(self: *Self) void {
+            if (mode == .Debug) {
```
Comment (alexrp) [inline lib/std/crypto/keccak_p.zig:297]: Other parts of the standard library generally enable safety checks like these in `ReleaseSafe` as well, FWIW.

**Pull request merged**